op-node/rollup/driver/state.go

```diff
@@ -462,35 +471,29 @@ type hashAndErrorChannel struct {
 	err  chan error
 }
 
-// checkForGapInUnsafeQueue checks if there is a gap in the unsafe queue and attempts to retrieve the missing payloads from the backup RPC.
-// WARNING: The sync client's attempt to retrieve the missing payloads is not guaranteed to succeed, and it will fail silently (besides
-// emitting warning logs) if the requests fail.
-func (s *Driver) checkForGapInUnsafeQueue(ctx context.Context) {
+// checkForGapInUnsafeQueue checks if there is a gap in the unsafe queue and attempts to retrieve the missing payloads from an alt-sync method.
+// WARNING: This is only an outgoing signal, the blocks are not guaranteed to be retrieved.
+// Results are received through OnUnsafeL2Payload.
+func (s *Driver) checkForGapInUnsafeQueue(ctx context.Context) error {
 	// subtract genesis time from wall clock to get the time elapsed since genesis, and then divide that
 	// difference by the block time to get the expected L2 block number at the current time. If the
 	// unsafe head does not have this block number, then there is a gap in the queue.
 	wallClock := uint64(time.Now().Unix())
 	genesisTimestamp := s.config.Genesis.L2Time
+	if wallClock < genesisTimestamp {
+		s.log.Debug("nothing to sync, did not reach genesis L2 time yet", "genesis", genesisTimestamp)
+		return nil
+	}
 	wallClockGenesisDiff := wallClock - genesisTimestamp
-	expectedL2Block := wallClockGenesisDiff / s.config.BlockTime
+	// Note: round down, we should not request blocks into the future.
+	blocksSinceGenesis := wallClockGenesisDiff / s.config.BlockTime
+	expectedL2Block := s.config.Genesis.L2.Number + blocksSinceGenesis
```

Comment:
Just being paranoid here:
1. The genesis block number is the point at which bedrock kicked in right? Or was there a consistent block time before bedrock as well?
2. Do we have appropriate throttling to handle the case where the sequencer is offline and so unsafe head stalls for a period? We don't want to spam our sync source for requests for blocks that don't exist yet.

Reply:
> The genesis block number is the point at which bedrock kicked in right? 

Yes. It's a bit of a weird name, "bedrock anchor point" is maybe more accurate. But for most op-stack chains, it'll be the same as genesis anyway.

Reply:
> Do we have appropriate throttling to handle the case where the sequencer is offline and so unsafe head stalls for a period?

Reply:
Looking at this again, at most we'd be scheduling a batch of requests every time the ticker goes off (2 * blockTime) which is not unreasonable.

Maybe worth logging a ticket to investigate further - it could be improved by detecting when a block we request is not found and bailing out rather than still sending a request for the whole range. It's not a new problem being introduced by this PR anyway.

Reply:
Tracked in CLI-3683 now